File: src/rulesEngine.js
const { cloneBoard, wouldBreakHive, canSlideBetween, removeTopPiece, isHiveConnected } = require('./board');
const { neighbors, key, directionIndex, directions, add } = require('./coords');
const { pieceDefinitions } = require('./pieces');
const { applyPlacement, applyMovement, nextPieceId } = require('./gameState');

const getPlayerPieces = (state, player) => {
  const pieces = [];
  for (const [coordKey, stack] of state.board.entries()) {
    if (!stack.length) continue;
    const top = stack[stack.length - 1];
    if (top.player !== player) continue;
    pieces.push({ coord: parseCoordKey(coordKey), stackHeight: stack.length, piece: top });
  }
  return pieces;
};

const parseCoordKey = (k) => {
  const [x, y, z] = k.split(',').map(Number);
  return { x, y, z };
};

const hivePiecesCount = (state) => {
  let count = 0;
  for (const stack of state.board.values()) {
    count += stack.length;
  }
  return count;
};

const hasQueenOnBoard = (state, player) =>
  Array.from(state.board.values()).some((stack) => stack.some((p) => p.player === player && p.type === 'queen'));

const queenRequirementActive = (state, player) => {
  const turnNumber = state.turnCounts[player] + 1; // turnCounts is completed turns
  return turnNumber >= 4 && !hasQueenOnBoard(state, player);
};

const emptyCellsAdjacentToBoard = (state) => {
  const empty = new Set();
  for (const [k] of state.board.entries()) {
    const coord = parseCoordKey(k);
    for (const n of neighbors(coord)) {
      const nk = key(n);
      if (!state.board.has(nk)) empty.add(nk);
    }
  }
  return Array.from(empty).map(parseCoordKey);
};

const cellsAdjacentToPlayer = (state, player) => {
  const empty = new Set();
  for (const [k, stack] of state.board.entries()) {
    if (!stack.some((p) => p.player === player)) continue;
    const coord = parseCoordKey(k);
    for (const n of neighbors(coord)) {
      const nk = key(n);
      if (!state.board.has(nk)) empty.add(nk);
    }
  }
  return Array.from(empty).map(parseCoordKey);
};

const placementMoves = (state, player) => {
  const mustPlaceQueen = queenRequirementActive(state, player);
  const availablePieces = Object.entries(state.reserves[player]).filter(
    ([type, count]) => count > 0 && (!mustPlaceQueen || type === 'queen'),
  );
  if (!availablePieces.length) return [];
  const placements = [];
  const isFirstMove = state.turnCounts[player] === 0 && hivePiecesCount(state) === 0;
  if (isFirstMove) {
    for (const [pieceType] of availablePieces) {
      placements.push({ type: 'place', pieceType, to: { x: 0, y: 0, z: 0 } });
    }
    return placements;
  }

  const potentialCells =
    state.turnCounts[player] === 0
      ? emptyCellsAdjacentToBoard(state)
      : cellsAdjacentToPlayer(state, player);

  for (const [pieceType] of availablePieces) {
    for (const cell of potentialCells) {
      const nk = key(cell);
      if (state.board.has(nk)) continue;
      if (state.turnCounts[player] > 0) {
        const neighborPieces = neighbors(cell)
          .map((c) => state.board.get(key(c)))
          .filter(Boolean)
          .flat();
        if (!neighborPieces.some((p) => p.player === player)) continue;
        if (neighborPieces.some((p) => p.player !== player)) continue;
      }
      placements.push({ type: 'place', pieceType, to: cell });
    }
  }
  return placements;
};

const touchesHive = (board, cell) => neighbors(cell).some((n) => board.has(key(n)));

const slidingNeighbors = (board, from) =>
  neighbors(from).filter(
    (n) => !board.has(key(n)) && canSlideBetween(board, from, n) && touchesHive(board, n),
  );

const spiderMoves = (state, from) => {
  const boardWithoutPiece = cloneBoard(state.board);
  removeTopPiece(boardWithoutPiece, from);
  const results = new Set();
  const dfs = (current, depth, visited) => {
    if (depth === 3) {
      results.add(key(current));
      return;
    }
    for (const n of slidingNeighbors(boardWithoutPiece, current)) {
      const nk = key(n);
      if (visited.has(nk)) continue;
      visited.add(nk);
      dfs(n, depth + 1, visited);
      visited.delete(nk);
    }
  };
  dfs(from, 0, new Set([key(from)]));
  return Array.from(results).map(parseCoordKey);
};

const antMoves = (state, from) => {
  const boardWithout = cloneBoard(state.board);
  removeTopPiece(boardWithout, from);
  const visited = new Set();
  const queue = [from];
  visited.add(key(from));
  const targets = new Set();
  while (queue.length) {
    const current = queue.shift();
    for (const n of slidingNeighbors(boardWithout, current)) {
      const nk = key(n);
      if (visited.has(nk)) continue;
      visited.add(nk);
      targets.add(nk);
      queue.push(n);
    }
  }
  targets.delete(key(from));
  return Array.from(targets).map(parseCoordKey);
};

const queenOrBeetleStepTargets = (state, from, allowsClimb) => {
  const targets = [];
  for (const n of neighbors(from)) {
    const nk = key(n);
    const occupied = state.board.has(nk);
    if (!occupied) {
      if (canSlideBetween(state.board, from, n)) targets.push(n);
    } else if (allowsClimb) {
      targets.push(n);
    }
  }
  return targets;
};

const grasshopperMoves = (state, from) => {
  const moves = [];
  for (const dir of directions) {
    let current = add(from, dir);
    let jumped = 0;
    while (true) {
      const k = key(current);
      if (!state.board.has(k)) {
        if (jumped > 0) moves.push(current);
        break;
      }
      jumped += 1;
      current = add(current, dir);
    }
  }
  return moves;
};

const generateMovesForPiece = (state, pieceEntry) => {
  const { coord, stackHeight, piece } = pieceEntry;
  const def = pieceDefinitions[piece.type];
  if (!def) return [];
  if (stackHeight > 1 && piece.type !== 'beetle') return [];
  if (wouldBreakHive(state.board, coord)) return [];
  const movement = def.movement;
  let targets = [];
  switch (movement.type) {
    case 'step': {
      targets = queenOrBeetleStepTargets(state, coord, movement.allowsClimb);
      break;
    }
    case 'exact_crawl': {
      targets = spiderMoves(state, coord);
      break;
    }
    case 'crawl_any_distance': {
      targets = antMoves(state, coord);
      break;
    }
    case 'jump_line': {
      targets = grasshopperMoves(state, coord);
      break;
    }
    default:
      break;
  }
  const filtered = targets
    .filter((t) => key(t) !== key(coord))
    .filter((t) => !wouldBreakHiveAfterMove(state, coord, t));
  return filtered.map((t) => ({
    type: 'move',
    pieceId: piece.id,
    from: coord,
    to: t,
  }));
};

const wouldBreakHiveAfterMove = (state, from, to) => {
  const boardCopy = cloneBoard(state.board);
  const moving = removeTopPiece(boardCopy, from);
  if (!moving) return true;
  const kTo = key(to);
  const stack = boardCopy.get(kTo) || [];
  boardCopy.set(kTo, [...stack, moving]);
  return !isHiveConnected(boardCopy);
};

const getLegalMoves = (state, player) => {
  if (state.currentPlayer !== player) return [];
  const mustPlaceQueen = queenRequirementActive(state, player);
  const moves = [];
  const placements = placementMoves(state, player);
  moves.push(...placements);
  if (mustPlaceQueen) {
    return dedupeMoves(moves);
  }
  const pieces = getPlayerPieces(state, player);
  for (const p of pieces) {
    moves.push(...generateMovesForPiece(state, p));
  }
  return dedupeMoves(moves);
};

const dedupeMoves = (moves) => {
  const seen = new Set();
  const uniq = [];
  for (const m of moves) {
    const sig = JSON.stringify(m);
    if (seen.has(sig)) continue;
    seen.add(sig);
    uniq.push(m);
  }
  return uniq;
};

const applyAction = (state, player, action) => {
  if (action.type === 'place') {
    const pieceId = nextPieceId(state, player, action.pieceType);
    return applyPlacement(state, player, action.pieceType, action.to, pieceId);
  }
  if (action.type === 'move') {
    return applyMovement(state, player, action.pieceId, action.from, action.to);
  }
  return state;
};

module.exports = {
  getLegalMoves,
  applyAction,
  queenRequirementActive,
  hivePiecesCount,
};

File: src/rules_engine.py
from __future__ import annotations

from typing import Dict, List, Set

from .board import (
    Board,
    can_slide_between,
    clone_board,
    is_hive_connected,
    remove_top_piece,
    top_piece,
    would_break_hive,
)
from .coords import Coord, DIRECTIONS, add, neighbors, parse_key
from .game_state import GameState, apply_movement, apply_placement, next_piece_id
from .pieces import piece_definitions


def _hive_piece_count(state: GameState) -> int:
    return sum(len(stack) for stack in state.board.values())


def _has_queen_on_board(state: GameState, player: str) -> bool:
    return any(piece["player"] == player and piece["type"] == "queen" for stack in state.board.values() for piece in stack)


def queen_requirement_active(state: GameState, player: str) -> bool:
    turn_number = state.turn_counts[player] + 1
    return turn_number >= 4 and not _has_queen_on_board(state, player)


def _empty_cells_adjacent(state: GameState) -> List[Coord]:
    empty = set()
    for k in state.board:
        coord = parse_key(k)
        for n in neighbors(coord):
            if n.key() not in state.board:
                empty.add(n.key())
    return [parse_key(k) for k in empty]


def _cells_adjacent_to_player(state: GameState, player: str) -> List[Coord]:
    empty = set()
    for k, stack in state.board.items():
        if not any(p["player"] == player for p in stack):
            continue
        coord = parse_key(k)
        for n in neighbors(coord):
            if n.key() not in state.board:
                empty.add(n.key())
    return [parse_key(k) for k in empty]


def placement_moves(state: GameState, player: str) -> List[dict]:
    must_place_queen = queen_requirement_active(state, player)
    available = [(t, c) for t, c in state.reserves[player].items() if c > 0 and (not must_place_queen or t == "queen")]
    if not available:
        return []
    placements: List[dict] = []
    is_first_move = state.turn_counts[player] == 0 and _hive_piece_count(state) == 0
    if is_first_move:
        for piece_type, _ in available:
            placements.append({"type": "place", "pieceType": piece_type, "to": Coord(0, 0, 0)})
        return placements

    potential_cells = _cells_adjacent_to_player(state, player) if state.turn_counts[player] > 0 else _empty_cells_adjacent(state)
    for piece_type, _ in available:
        for cell in potential_cells:
            if cell.key() in state.board:
                continue
            neighbor_pieces = [
                piece for n in neighbors(cell) for piece in state.board.get(n.key(), [])
            ]
            if state.turn_counts[player] > 0:
                if not any(p["player"] == player for p in neighbor_pieces):
                    continue
                if any(p["player"] != player for p in neighbor_pieces):
                    continue
            placements.append({"type": "place", "pieceType": piece_type, "to": cell})
    return placements


def _touches_hive(board: Board, cell: Coord) -> bool:
    return any(n.key() in board for n in neighbors(cell))


def _sliding_neighbors(board: Board, from_coord: Coord) -> List[Coord]:
    return [
        n
        for n in neighbors(from_coord)
        if n.key() not in board and can_slide_between(board, from_coord, n) and _touches_hive(board, n)
    ]


def _spider_moves(state: GameState, from_coord: Coord) -> List[Coord]:
    board_without = clone_board(state.board)
    remove_top_piece(board_without, from_coord)
    results = set()

    def dfs(current: Coord, depth: int, visited: Set[str]) -> None:
        if depth == 3:
            results.add(current.key())
            return
        for n in _sliding_neighbors(board_without, current):
            nk = n.key()
            if nk in visited:
                continue
            visited.add(nk)
            dfs(n, depth + 1, visited)
            visited.remove(nk)

    dfs(from_coord, 0, {from_coord.key()})
    return [parse_key(k) for k in results]


def _ant_moves(state: GameState, from_coord: Coord) -> List[Coord]:
    board_without = clone_board(state.board)
    remove_top_piece(board_without, from_coord)
    visited = {from_coord.key()}
    queue = [from_coord]
    targets = set()
    while queue:
        current = queue.pop(0)
        for n in _sliding_neighbors(board_without, current):
            nk = n.key()
            if nk in visited:
                continue
            visited.add(nk)
            targets.add(nk)
            queue.append(n)
    targets.discard(from_coord.key())
    return [parse_key(k) for k in targets]


def _step_targets(state: GameState, from_coord: Coord, allows_climb: bool) -> List[Coord]:
    targets: List[Coord] = []
    for n in neighbors(from_coord):
        occupied = n.key() in state.board
        if not occupied:
            if can_slide_between(state.board, from_coord, n):
                targets.append(n)
        elif allows_climb:
            targets.append(n)
    return targets


def _grasshopper_moves(state: GameState, from_coord: Coord) -> List[Coord]:
    moves: List[Coord] = []
    for direction in DIRECTIONS:
        current = add(from_coord, direction)
        jumped = 0
        while True:
            if current.key() not in state.board:
                if jumped > 0:
                    moves.append(current)
                break
            jumped += 1
            current = add(current, direction)
    return moves


def _generate_moves_for_piece(state: GameState, coord: Coord, stack_height: int, piece: dict) -> List[dict]:
    definition = piece_definitions.get(piece["type"])
    if not definition:
        return []
    if stack_height > 1 and piece["type"] != "beetle":
        return []
    if would_break_hive(state.board, coord):
        return []

    movement = definition["movement"]
    targets: List[Coord] = []
    if movement["type"] == "step":
        targets = _step_targets(state, coord, movement.get("allowsClimb", False))
    elif movement["type"] == "exact_crawl":
        targets = _spider_moves(state, coord)
    elif movement["type"] == "crawl_any_distance":
        targets = _ant_moves(state, coord)
    elif movement["type"] == "jump_line":
        targets = _grasshopper_moves(state, coord)

    filtered = [
        t
        for t in targets
        if t.key() != coord.key() and not _would_break_hive_after_move(state, coord, t)
    ]
    return [{"type": "move", "pieceId": piece["id"], "from": coord, "to": t} for t in filtered]


def _would_break_hive_after_move(state: GameState, from_coord: Coord, to_coord: Coord) -> bool:
    board_copy = clone_board(state.board)
    moving = remove_top_piece(board_copy, from_coord)
    if not moving:
        return True
    board_copy.setdefault(to_coord.key(), []).append(moving)
    return not is_hive_connected(board_copy)


def _player_pieces(state: GameState, player: str) -> List[dict]:
    pieces = []
    for coord_key, stack in state.board.items():
        if not stack:
            continue
        top = stack[-1]
        if top["player"] != player:
            continue
        pieces.append({"coord": parse_key(coord_key), "stackHeight": len(stack), "piece": top})
    return pieces


def _move_signature(move: dict) -> tuple:
    from_key = move.get("from").key() if move.get("from") else None
    to_key = move.get("to").key() if move.get("to") else None
    return (move.get("type"), move.get("pieceId"), move.get("pieceType"), from_key, to_key)


def _dedupe_moves(moves: List[dict]) -> List[dict]:
    seen: Set[tuple] = set()
    unique = []
    for m in moves:
        sig = _move_signature(m)
        if sig in seen:
            continue
        seen.add(sig)
        unique.append(m)
    return unique


def get_legal_moves(state: GameState, player: str) -> List[dict]:
    if state.current_player != player:
        return []
    must_place_queen = queen_requirement_active(state, player)
    moves: List[dict] = placement_moves(state, player)
    if must_place_queen:
        return _dedupe_moves(moves)
    for entry in _player_pieces(state, player):
        moves.extend(_generate_moves_for_piece(state, entry["coord"], entry["stackHeight"], entry["piece"]))
    return _dedupe_moves(moves)


def apply_action(state: GameState, player: str, action: dict) -> GameState:
    if action["type"] == "place":
        piece_id = action.get("pieceId") or next_piece_id(state, player, action["pieceType"])
        return apply_placement(state, player, action["pieceType"], action["to"], piece_id)
    if action["type"] == "move":
        return apply_movement(state, player, action["pieceId"], action["from"], action["to"])
    return state
